File: examples/orchestration/python/01_basic_flow.py
#!/usr/bin/env python3
"""Basic Flow Execution

Demonstrates the minimal engine setup:
1. Create an EngineBuilder
2. Point it at a flow store directory containing JSON flow definitions
3. Build the engine
4. Execute a flow and collect all events
"""

import os
import shutil
import tempfile

from cleargate import EngineBuilder


def main():
    # Copy the echo flow into a temp directory so the engine can find it.
    flow_src = os.path.join(os.path.dirname(__file__), "..", "flows", "echo-flow.json")
    with tempfile.TemporaryDirectory() as tmpdir:
        flow_dir = os.path.join(tmpdir, "flows")
        run_dir = os.path.join(tmpdir, "runs")
        os.makedirs(flow_dir, exist_ok=True)

        # The engine expects flows in <flow_store>/<flow_id>/flow.json
        flow_dest = os.path.join(flow_dir, "echo-flow")
        os.makedirs(flow_dest, exist_ok=True)
        shutil.copy(flow_src, os.path.join(flow_dest, "flow.json"))

        # Build the engine
        builder = EngineBuilder()
        builder.flow_store_path(flow_dir)
        builder.run_store_path(run_dir)
        builder.crash_recovery(False)
        engine = builder.build()

        print("Engine built. Registered nodes:")
        catalog = engine.node_catalog()
        for node in catalog:
            print(f"  - {node['node_type']} ({node['category']})")

        # Execute the echo flow
        print("\nExecuting echo-flow...")
        handle = engine.execute("echo-flow", {"message": "Hello from Python!"})
        print(f"Run ID: {handle.run_id}")

        # Collect all events
        events = handle.wait()
        print(f"\nReceived {len(events)} events:")
        for event in events:
            # WriteEvent is a tagged enum — the type is the key
            if isinstance(event, dict):
                for key in event:
                    print(f"  [{key}]")
                    break

        engine.shutdown()
        print("\nDone!")


if __name__ == "__main__":
    main()

File: examples/orchestration/node/src/01_basic_flow.ts
/**
 * Basic Flow Execution
 *
 * Demonstrates the minimal engine setup:
 * 1. Create an EngineBuilder
 * 2. Point it at flow/run store directories
 * 3. Build the engine
 * 4. Execute a flow and collect all events
 */

import * as fs from "fs";
import * as path from "path";
import * as os from "os";

// The native binding is loaded via require() by the SDK
const native = require("../../../crates/cleargate-node/cleargate.node");

function main() {
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), "cleargate-"));
  const flowDir = path.join(tmpDir, "flows");
  const runDir = path.join(tmpDir, "runs");

  // Copy echo flow
  const flowDest = path.join(flowDir, "echo-flow");
  fs.mkdirSync(flowDest, { recursive: true });
  fs.copyFileSync(
    path.join(__dirname, "../../flows/echo-flow.json"),
    path.join(flowDest, "flow.json"),
  );

  try {
    const builder = new native.NapiEngineBuilder();
    builder.flowStorePath(flowDir);
    builder.runStorePath(runDir);
    builder.crashRecovery(false);
    const engine = builder.build();

    console.log("Engine built. Node catalog:");
    const catalog = engine.nodeCatalog();
    for (const node of catalog) {
      console.log(`  - ${node.node_type} (${node.category})`);
    }

    console.log("\nExecuting echo-flow...");
    const handle = engine.execute("echo-flow", {
      message: "Hello from Node.js!",
    });
    console.log(`Run ID: ${handle.runId}`);

    const events = handle.waitForCompletion();
    console.log(`\nReceived ${events.length} events:`);
    for (const event of events) {
      const type = Object.keys(event)[0];
      console.log(`  [${type}]`);
    }

    engine.shutdown();
    console.log("\nDone!");
  } finally {
    fs.rmSync(tmpDir, { recursive: true, force: true });
  }
}

main();
